src/webgpu/api/validation/copyTextureToTexture.spec.ts

```diff
@@ -274,157 +281,311 @@ g.test('depth_stencil_copy_restrictions')
     });
 
+    const srcSizeAtLevel = t.GetPhysicalSubresourceSize(srcTextureSize, srcCopyLevel);
+    const dstSizeAtLevel = t.GetPhysicalSubresourceSize(dstTextureSize, dstCopyLevel);
+
+    const copyOrigin = { x: copyBoxOffsets.x, y: copyBoxOffsets.y, z: 0 };
+
+    const copyWidth =
+      Math.min(srcSizeAtLevel.width, dstSizeAtLevel.width) + copyBoxOffsets.width - copyOrigin.x;
+    const copyHeight =
+      Math.min(srcSizeAtLevel.height, dstSizeAtLevel.height) + copyBoxOffsets.height - copyOrigin.y;
+
     const isSuccess =
-      copyWidth === srcTextureSize >> srcCopyLevel &&
-      copyHeight === srcTextureSize >> srcCopyLevel &&
-      copyWidth === dstTextureSize >> dstCopyLevel &&
-      copyHeight === dstTextureSize >> dstCopyLevel;
+      copyWidth === srcSizeAtLevel.width &&
+      copyHeight === srcSizeAtLevel.height &&
+      copyWidth === dstSizeAtLevel.width &&
+      copyHeight === dstSizeAtLevel.height;
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture, origin: { x: 0, y: 0, z: 0 }, mipLevel: srcCopyLevel },
+      { texture: dstTexture, origin: copyOrigin, mipLevel: dstCopyLevel },
+      { width: copyWidth, height: copyHeight, depth: 1 },
+      isSuccess
+    );
     t.TestCopyTextureToTexture(
-      { texture: srcTexture, origin: srcCopyOrigin, mipLevel: srcCopyLevel },
-      { texture: dstTexture, origin: dstCopyOrigin, mipLevel: dstCopyLevel },
+      { texture: srcTexture, origin: copyOrigin, mipLevel: srcCopyLevel },
+      { texture: dstTexture, origin: { x: 0, y: 0, z: 0 }, mipLevel: dstCopyLevel },
       { width: copyWidth, height: copyHeight, depth: 1 },
       isSuccess
     );
   });
 
 g.test('copy_ranges')
-  .params(
-    params()
-      .combine(
-        poptions('srcCopyOrigin', [
-          { x: 0, y: 0, z: 0 },
-          { x: 0, y: 0, z: 1 },
-          { x: 0, y: 0, z: 2 },
-          { x: 8, y: 4, z: 0 },
-          { x: 8, y: 4, z: 1 },
-          { x: 16, y: 8, z: 0 },
-          { x: 16, y: 8, z: 1 },
-        ])
-      )
-      .combine(
-        poptions('dstCopyOrigin', [
-          { x: 0, y: 0, z: 0 },
-          { x: 0, y: 0, z: 1 },
-          { x: 0, y: 0, z: 2 },
-          { x: 8, y: 4, z: 0 },
-          { x: 8, y: 4, z: 1 },
-          { x: 16, y: 8, z: 0 },
-          { x: 16, y: 8, z: 1 },
-        ])
-      )
-      .combine(
-        poptions('copyExtent', [
-          { width: 0, height: 0, depth: 0 },
-          { width: 0, height: 0, depth: 1 },
-          { width: 0, height: 0, depth: 2 },
-          { width: 8, height: 0, depth: 1 },
-          { width: 0, height: 4, depth: 1 },
-          { width: 8, height: 4, depth: 1 },
-          { width: 8, height: 4, depth: 2 },
-          { width: 16, height: 8, depth: 1 },
-          { width: 16, height: 8, depth: 2 },
-          { width: 16, height: 16, depth: 0 },
-        ])
-      )
-      .combine(poptions('srcCopyLevel', [0, 1]))
-      .combine(poptions('dstCopyLevel', [0, 1]))
-  )
+  .params([
```

Comment:
This big list is ok, but consider:
```ts
g.test()
  .params([
    ...params()
      .combine(poptions('copyBoxOffsets', [
        { x: 0, y: 0, z: 0, width: 0, height: 0, depth: -2 },
        // ...
      ]))
      .combine(poptions('srcCopyLevel', [0, 1]))
      .combine(poptions('dstCopyLevel', [0, 1])),
    ...params()
      .combine(poptions('copyBoxOffsets', [
        { x: 0, y: 0, z: 0, width: 1, height: 0, depth: -2 },
        // ...
      ]))
      .combine([{ srcCopyLevel: 0, dstCopyLevel: 0 }])
  ] as const)
```

(Let me know if TypeScript has issues with doing this, I may need to fix something. You may also need more `as const` throughout.)